Fixes #18345 - Improve templates error handling

ares opened the pull request.

ares committed:
Fixes #18345 - Improve templates error handling

ares commented:
Before

![error1](https://cloud.githubusercontent.com/assets/109773/22551462/368cfdf4-e955-11e6-8dc0-267c69a954a7.png)

the trace does not tell anything about where the error occurred

```
 | Foreman::Exception: ERF42-1699 [Foreman::Exception]: The snippet 'kickstart_networking_setup' threw an error: ERF42-7124 [Foreman::Exception]: The snippet 'kickstart_ifcfg_bond_interface' threw an error: ERF22-8399 [Foreman::Renderer:
:SyntaxError]: Syntax error occurred on parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: (erb):11: syntax error, unexpected ';'
 | ;  1 + ; _erbout.force_encoding(__ENCODING__)
 |         ^
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `eval'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `result'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:63:in `render_safe'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:165:in `unattended_render'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:114:in `snippet'
 | (erb):6:in `block in render_safe'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activerecord-4.2.7.1/lib/active_record/relation/delegation.rb:46:in `each'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activerecord-4.2.7.1/lib/active_record/relation/delegation.rb:46:in `each'
 | (erb):3:in `render_safe'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `eval'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `result'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:63:in `render_safe'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:165:in `unattended_render'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:114:in `snippet'
 | (erb):151:in `render_safe'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `eval'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `result'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:63:in `render_safe'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:165:in `unattended_render'
 | /home/ares/Projekty/Zdrojaky/foreman/app/controllers/templates_controller.rb:106:in `safe_render'
 | /home/ares/Projekty/Zdrojaky/foreman/app/controllers/templates_controller.rb:95:in `preview'
```

After

![error3](https://cloud.githubusercontent.com/assets/109773/22551457/2dc713c6-e955-11e6-9c4f-0c0ef44dd199.png)

the trace contains template names (including snippets)
```
| /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `eval'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `result'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:63:in `render_safe'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:165:in `unattended_render'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:114:in `snippet'
 | kickstart_networking_setup:6:in `block in render_safe'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activerecord-4.2.7.1/lib/active_record/relation/delegation.rb:46:in `each'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/activerecord-4.2.7.1/lib/active_record/relation/delegation.rb:46:in `each'
 | kickstart_networking_setup:3:in `render_safe'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `eval'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `result'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:63:in `render_safe'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:165:in `unattended_render'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:114:in `snippet'
 | Kickstart default:151:in `render_safe'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `eval'
 | /home/ares/.rbenv/versions/2.2.2/lib/ruby/2.2.0/erb.rb:863:in `result'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:63:in `render_safe'
 | /home/ares/Projekty/Zdrojaky/foreman/lib/foreman/renderer.rb:165:in `unattended_render'
 | /home/ares/Projekty/Zdrojaky/foreman/app/controllers/templates_controller.rb:106:in `safe_render'
 | /home/ares/Projekty/Zdrojaky/foreman/app/controllers/templates_controller.rb:95:in `preview'
```

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -48,12 +51,23 @@ def host_param_false?(name)
+
+    rescue ::Racc::ParseError, ::SyntaxError => e
+      # Racc::ParseError is raised in safe mode, SyntaxError in unsafe mode
+      new_e = Foreman::Renderer::SyntaxError.new(_("Syntax error occurred on parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: ") + e.message)
```
1. Untranslated strings (N_()) must be given as the first parameter to a Foreman::Exception, interpolated values passed in as additional arguments.
1. "occurred _while_ parsing the template"

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -99,7 +113,10 @@ def snippet(name, options = {})
         rescue => exc
-          raise "The snippet '#{name}' threw an error: #{exc}"
+          # fixes the backtrace but ideally should raise better error message with Foreman rendering exception, same everywhere in this file
+          e = ::Foreman::Exception.new "The snippet '#{name}' threw an error: #{exc}"
```
Missing correct string extraction for Foreman::Exception.

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -99,7 +113,10 @@ def snippet(name, options = {})
           return unattended_render(template, nil, options[:variables] || {})
         rescue => exc
-          raise "The snippet '#{name}' threw an error: #{exc}"
+          # fixes the backtrace but ideally should raise better error message with Foreman rendering exception, same everywhere in this file
```
Not sure that I understand this comment, does it mean it should use RenderingException, or provide different information?

ares replied:
Yes, you're right. I didn't want to do a big refactoring of exceptions in the renderer in this commit but when we do it one day, this should have it's own exception class. If you're OK with introducing SnippetError for this case within this PR, I'm happy to do it. Or I can also just remove the comment, whatever you prefer.

domcleal replied:
Either using a different exception or leaving it as-is (but minus the comment, which I don't think is really needed) is fine, provided that the current exception captures the original message. If using a different exception class then it should probably derive from Foreman::WrappedException?

ares replied:
Hm, the WrappedException hierarchy does not work in here, I want all exceptions raised by renderer have one commont parent so it's easy to rescue them only. Not every renderer exception will wrap another exception, WrappedException should have been concern probably. Anyway I'll keep exceptions refactoring for later, I removed the comment.

domcleal commented on test/unit/foreman/renderer_test.rb:
```diff
@@ -97,6 +97,13 @@ def setup_safemode_renderer
 
+    test "#{renderer_name} should raise renderer syntax error on syntax error" do
+      send "setup_#{renderer_name}"
+      assert_raises Foreman::Renderer::SyntaxError do
```
Can you add assertions to check the exception's correct? (`e = assert_raises...`)

Particularly that the backtrace contains the template name and the message has correctly wrapped the original syntax error.

domcleal reviewed (changes requested):
Broadly fine, but check correct usage of strings and Foreman::Exception (subclasses).

ares commented:
Thanks @domcleal, I think I addressed your comments except one, please take another look and let me know about your preferences.

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -48,20 +51,30 @@ def host_param_false?(name)
+
+    rescue ::Racc::ParseError, ::SyntaxError => e
+      # Racc::ParseError is raised in safe mode, SyntaxError in unsafe mode
+      new_e = Foreman::Renderer::SyntaxError.new(N_("Syntax error occurred while parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: ") + e.message)
```
Use %s in the message and pass `e.message` in as the second arg to the constructor.

ares commented:
Thanks @domcleal, I fixed both comments and also modified the test that verifies line number and template name to only run if we have recent ERB (ruby 2.2+)

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -107,7 +120,9 @@ def snippet(name, options = {})
           return unattended_render(template, nil, options[:variables] || {})
         rescue => exc
-          raise "The snippet '#{name}' threw an error: #{exc}"
+          e = ::Foreman::Exception.new(N_("The snippet '%{name}' threw an error: %{exc}"), { :name => name, :exc => exc })
```
With the `rescue` in `unattended_render`, I think this rescue block is obsolete. If you remove it then it'll reduce the length and complexity of the error message:

    «ERF42-7327 [Foreman::Exception]: »The snippet 'kickstart_networking_setup' threw an error: ERF22-8069 [Foreman::Renderer::SyntaxError]: »Syntax error occurred while parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: kickstart_networking_setup:2: syntax error, unexpected [..]

Without this rescue block then it could simply be:

    ERF22-8069 [Foreman::Renderer::SyntaxError]: »Syntax error occurred while parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: kickstart_networking_setup:2: syntax error, unexpected [..]

which removes the unnecessary second error code and duplication in the message.

ares replied:
I'm sorry I fail to locate the `rescue` in `unattended_render` you mention, could you please point me to it so I could better understand the change? If you meant the nested in in `safe_render` then it only catches syntax errors. If that would be other exception, e.g. NoMethodError when nil does not respond to something, it would not be rescued properly. Maybe we could skip the wrapping of exception and simply re-raise it if `exc` is kind of `Foreman::Exception`, in such case we know it was already handled by us.

domcleal commented on lib/foreman/renderer.rb:
```diff
@@ -48,20 +51,30 @@ def host_param_false?(name)
+
+    rescue ::Racc::ParseError, ::SyntaxError => e
+      # Racc::ParseError is raised in safe mode, SyntaxError in unsafe mode
+      new_e = Foreman::Renderer::SyntaxError.new(N_("Syntax error occurred while parsing the template, make sure you have all ERB tags properly closed and the Ruby syntax is valid. The Ruby error: %s"), e.message)
```
Add `template_name` into this message, for when the name isn't set via `location=`? (particularly if you remove the `rescue` block from `snippet` as suggested)

domcleal reviewed (changes requested):
Thanks, works well. I'd like to see the snippet syntax error message shortened though, it's become much longer.

ares commented:
@domcleal I updated the PR. The second comment is addresses as I suggested, in my testing seems good also for NoMethodError-like cases. Please take another look.

domcleal commented:
Thanks, looks fine. Could you please rebase the PR to check the tests pass on Ruby 2.4?

ares commented:
rebased, fingers crossed

domcleal commented:
Merged, thanks @ares.
